File: 2025/python/day03.py
from utils import get_line_data
from utils import timeit


def highest_possible_joltage(line: str, n: int = 2) -> int:
    digits = list(line[:n])

    for i in range(n, len(line)):
        digits.append(line[i])

        best_number = 0
        best_digits = None

        for j in range(len(digits)):
            candidate = digits[:j] + digits[j+1:]
            number = int("".join(candidate))
            if number > best_number:
                best_number = number
                best_digits = candidate

        digits = best_digits

    return int("".join(digits))


@timeit
def main():
    data = get_line_data("03")

    print(sum(highest_possible_joltage(line) for line in data))
    print(sum(highest_possible_joltage(line, 12) for line in (data)))

File: 2025/rust/day03.rs
use crate::y2025::utils;

fn highest_possible_joltage(line: &str, n: usize) -> u64 {
    let mut digits: Vec<char> = line.chars().take(n).collect();

    for i in n..line.len() {
        digits.push(line.chars().nth(i).unwrap());

        let mut best_number = 0u64;
        let mut best_digits: Option<Vec<char>> = None;

        for j in 0..digits.len() {
            let mut candidate = digits.clone();
            candidate.remove(j);
            let number_str: String = candidate.iter().collect();
            if let Ok(number) = number_str.parse::<u64>() {
                if number > best_number {
                    best_number = number;
                    best_digits = Some(candidate);
                }
            }
        }

        if let Some(best) = best_digits {
            digits = best;
        }
    }

    let result_str: String = digits.iter().collect();
    result_str.parse::<u64>().unwrap_or(0)
}

pub fn main() {
    let data = utils::get_line_data("03");

    let p1: u64 = data.iter().map(|line| highest_possible_joltage(line, 2)).sum();
    let p2: u64 = data.iter().map(|line| highest_possible_joltage(line, 12)).sum();

    println!("{}", p1);
    println!("{}", p2);
}
